promql/functions.go

```diff
@@ -184,6 +184,137 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
 	return resultVector
 }
 
+// Calculate the smoothed value at index i in raw data d.
+// The smoothed value is created by creating two values (x,y) where...
+// x = the current raw data point, weighted by the smoothing factor.
+// y = the last smoothed point + the last trend value weighted by the inverse of the smoothing factor.
+// s = computed smoothed values, b = computed trend factors, d = raw
+func doubleSVal(i int, sf, tf float64, s, b, d []float64) float64 {
+
+	// check for the cached value
```

Comment:
Having to track this seems odd for dynamic programming, you should already know what has been calculated.

For example the `doubleSVal(i-1)` below can be a straight lookup, and if you special case on `i` you should be able to remove the rest.